configs/deprecated/example/se.py

```diff
@@ -119,11 +107,92 @@ def get_processes(args):
         return multiprocesses, 1
 
 
+#
+# Following constants are taken from elf.h.
+#
+ELFMAG = b"\177ELF"
+
+EI_DATA = 5  #  Data encoding byte index
+ELFDATANONE = 0  #  Invalid data encoding
+ELFDATA2LSB = 1  #  2's complement, little endian
+ELFDATA2MSB = 2  #  2's complement, big endian
+
+# Legal values for e_machine (architecture).
+EM_NONE = 0  #  No machine
+EM_SPARC = 2  #  SUN SPARC
+EM_386 = 3  #  Intel 80386
+EM_MIPS = 8  #  MIPS R3000 big-endian
+EM_MIPS_RS3_LE = 10  #  MIPS R3000 little-endian
+EM_PPC = 20  #  PowerPC
+EM_PPC64 = 21  #  PowerPC 64-bit
+EM_ARM = 40  #  ARM
+EM_X86_64 = 62  #  AMD x86-64 architecture
+EM_AARCH64 = 183  #  ARM AARCH64
+EM_AMDGPU = 224  #  AMD GPU
+EM_RISCV = 243  #  RISC-V
+
+# Non-standard
+EI_E_MACHINE = 18  # e_machine first byte index
+
+
+def set_isa(args):
+    """
+    Set the architecture based on executable (if this gem5 has
+    support for multiple architectures).
+    """
+    if len(get_supported_isas()) > 1:
+        exe_path = args.cmd.split(";")[0]
```

Comment:
In your code you are setting the TARGET_ISA based on the first binary passed as an argument. What happens if you have multiple binaries with different ISAs ? Basically you will be using the same data structures (like PageTableWalkerCache) for all cpus (get_runtime_isa will always return the same value), while you should make distinction between them.

As I commented earlier, I believe we should rewrite the common module to make it multi-ISA compatible (for example extracting the ISA from the CPU type or the cpu.isa object....)

Reply:
> In that case you'd get either illegal instruction or even load error for the firts binary that's for other ISA than the first.

Not if the simulation really assumes multiple CPUs with different ISAs
So I can have an Arm CPU wunning an Arm workloads and a RISCV CPU running a RISCV workload in the same simulation. If things are done properly no error will arise.

> Was this scenario supported back in the days se.py was not deprecated and there was no multi-ISA gem5?

Yes without multi-ISA, the non TARGET_ISA binary will fail to be loaded by the ElfLoader class (in C++ https://github.com/gem5/gem5/blob/stable/src/base/loader/elf_object.cc#L222) when comparing the Arch with the Workloads class (https://github.com/gem5/gem5/blob/stable/src/arch/arm/linux/se_workload.cc#L67)